src/transformStateWithClones.js

```diff
@@ -7,7 +7,29 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
-}
+  const stateHistory = [];
+  const newState = { ...state };
+
+  for (const action of actions) {
+    if (action.type === 'addProperties') {
+      Object.assign(newState, action.extraData);
+    }
+
+    if (action.type === 'removeProperties') {
+      for (const key of action.keysToRemove) {
+        delete newState[key];
+      }
+    }
 
+    if (action.type === 'clear') {
+      for (const key in newState) {
+        delete newState[key];
+      }
+    }
```

Comment:
Each case within the `switch` statement should have its own scope for the new state to prevent unintended mutations to the state from previous actions. This can be achieved by creating a new state copy within each case.